### ISSUE TITLE
MPAS-seaice OpenMP gnu compile error

### ISSUE THREAD

**Issue by mark-petersen** [MPAS-Dev/MPAS-Model#286]
MPAS seaice compilation with gnu and OpenMP dies on `!$OMP declare simd` lines with current head, 8cb43ea1. Compile error does not occur with intel or with gnu when `OPENMP=false`. Compile error occurs with MPAS-seaice standalone on grizzly, and E3SM compile on cori-knl.

**Comment by mark-petersen** [comment]
MPAS compilation on LANL-IC:
```
module load gcc/5.3.0 openmpi/1.10.5 netcdf/4.4.1 parallel-netcdf/1.5.0
make gfortran CORE=seaice OPENMP=true
```
dies with:
```
mpas_seaice_velocity_solver_constitutive_relation.F:161:51:

 !$omp declare simd(seaice_evp_constitutive_relation)
                                                   1
Error: Symbol ‘seaice_evp_constitutive_relation’ at (1) has already been host associated
mpas_seaice_velocity_solver_constitutive_relation.F:241:59:

 !$omp declare simd(seaice_evp_constitutive_relation_revised)
                                                           1
Error: Symbol ‘seaice_evp_constitutive_relation_revised’ at (1) has already been host associated
mpas_seaice_velocity_solver_constitutive_relation.F:317:54:

 !$omp declare simd(seaice_linear_constitutive_relation)
                                                      1
Error: Symbol ‘seaice_linear_constitutive_relation’ at (1) has already been host associated
```

**Comment by mark-petersen** [comment]
E3SM compilation, cori-knl, current master 3c427c2dd:

<details>
<summary> E3SM error text</summary> 

```
cd cime/scripts/
./create_test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu -p acme -q debug --walltime 00:30:00
Creating test directory /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.20190618_144616_8rok4p
RUNNING TESTS:
  SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu
Starting CREATE_NEWCASE for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu with 1 procs
Finished CREATE_NEWCASE for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu in 2.322259 seconds (PASS)
Starting XML for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu with 1 procs
Finished XML for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu in 0.335063 seconds (PASS)
Starting SETUP for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu with 1 procs
Finished SETUP for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu in 6.831086 seconds (PASS)
Starting SHAREDLIB_BUILD for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu with 1 procs
Finished SHAREDLIB_BUILD for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu in 370.683789 seconds (PASS)
Starting MODEL_BUILD for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu with 4 procs
Finished MODEL_BUILD for test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu in 354.278317 seconds (FAIL). [COMPLETED 1 of 1]
    Case dir: /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.20190618_144616_8rok4p
    Errors were:
        Building test for SMS in directory /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.20190618_144616_8rok4p
        Error: Symbol 'seaice_evp_constitutive_relation' at (1) has already been host associated

        Error: Symbol 'seaice_evp_constitutive_relation_revised' at (1) has already been host associated

        Error: Symbol 'seaice_linear_constitutive_relation' at (1) has already been host associated

        Error: Symbol 'seaice_evp_constitutive_relation' at (1) has already been host associated

        Error: Symbol 'seaice_evp_constitutive_relation_revised' at (1) has already been host associated

        Error: Symbol 'seaice_linear_constitutive_relation' at (1) has already been host associated

        ERROR: BUILD FAIL: mpassi.buildlib failed, cat /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.20190618_144616_8rok4p/bld/ice.bldlog.190618-145239

Due to presence of batch system, create_test will exit before tests are complete.
To force create_test to wait for full completion, use --wait
At test-scheduler close, state is:
FAIL SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu (phase MODEL_BUILD)
    Case dir: /global/cscratch1/sd/mpeterse/acme_scratch/cori-knl/SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu.20190618_144616_8rok4p
test-scheduler took 735.790168047 seconds
```
</details>

**Comment by mark-petersen** [comment]
When this is fixed, also test with intel with debug and openmp:
```
./create_test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel -p acme -q debug --walltime 00:30:00
```

**Comment by amametjanov** [comment]
Can you upgrade the gnu compiler?

**Comment by mark-petersen** [comment]
Same error on LANL IC with `gcc/6.4.0`. I don't have pnetcdf & pio compiled for gcc v7 right now. But we get the same error on cori-knl with gnu. @amametjanov do you get a clean compile with a different gcc compiler?

**Comment by amametjanov** [comment]
GNU v.7.3.0 or later should work and also provide better support for OpenMP.


On Wed, Jun 19, 2019 at 12:50 PM Mark Petersen <notifications@github.com<mailto:notifications@github.com>> wrote:

Same error on LANL IC with gcc/6.4.0. I don't have pnetcdf & pio compiled for gcc v7 right now. But we get the same error on cori-knl with gnu. @amametjanov<https://github.com/amametjanov> do you get a clean compile with a different gcc compiler?

**Comment by mark-petersen** [comment]
Strange. cori-knl and theta both use gnu 7.3.0. cori-knl fails compile on this test but theta does not:
```
./create_test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu -p acme -q debug --walltime 00:30:00
./create_test SMS_D_Ln9.T62_oQU120_ais20.MPAS_LISIO_TEST.theta_gnu -p OceanClimate_2
```

I'm looking in 
```
cime/config/e3sm/machines/config_machines.xml
 343   <machine MACH="cori-knl">
...
 438       <modules compiler="gnu">
 439         <command name="swap">PrgEnv-intel PrgEnv-gnu/6.0.4</command>
 440         <command name="rm">gcc</command>
 441         <command name="load">gcc/7.3.0</command>
 442         <command name="rm">cray-libsci</command>
 443         <command name="load">cray-libsci/18.03.1</command>
 444       </modules>
```
theta and cori-knl both use gcc/7.3.0. There must be some other difference. Do they both use OpenMP in the same way?  I see this for theta:
```
1665     <environment_variables SMP_PRESENT="TRUE" compiler="intel">
1666       <env name="SMP_VARS">-e OMP_NUM_THREADS=$ENV{OMP_NUM_THREADS} -e OMP_STACKSIZE=128M -e KMP_AFFINITY=granularity=thread,scatter</env>
1667     </environment_variables>
1668     <environment_variables SMP_PRESENT="TRUE" compiler="!intel">
1669       <env name="SMP_VARS">-e OMP_NUM_THREADS=$ENV{OMP_NUM_THREADS} -e OMP_STACKSIZE=128M -e OMP_PROC_BIND=spread -e OMP_PLACES=threads</env>
1670     </environment_variables>
```
and this for cori-knl:
```
 482     <environment_variables>
 483       <env name="MPICH_ENV_DISPLAY">1</env>
 484       <env name="MPICH_VERSION_DISPLAY">1</env>
 485       <!--env name="MPICH_CPUMASK_DISPLAY">1</env-->
 486
 487       <env name="OMP_STACKSIZE">128M</env>
 488       <env name="OMP_PROC_BIND">spread</env>
 489       <env name="OMP_PLACES">threads</env>
 490       <env name="HDF5_USE_FILE_LOCKING">FALSE</env>
 491     </environment_variables>
```
but it looks like the openmpi version is hidden in the programming environment.

Whatever it is, we need to be able to run with the gnu compiler on cori-knl. Maybe I should make an E3SM ticket instead.

**Comment by amametjanov** [comment]
Fixed by MPAS-Dev/MPAS-Model#348. "Close issue" button is disabled for me. @jonbob or @mark-petersen, can you please close? Thanks.

**Comment by jonbob** [comment]
Got it, @amametjanov. Thanks for taking care of this
